### PR TITLE
Add a title for custom title bar style

### PR DESCRIPTION
**Pull request by msujew**
#### What it does

Closes https://github.com/eclipse-theia/theia/issues/7076

When selecting the `custom` title bar style, the application currently shows no title. This change adresses the issue by trying to replicate what VSCode does with its title bars. Also aligns the title behavior of Theia to VSCode which shows the currently opened workspace and active main area widget in the title. For this, some more title-based infrastructure has been introduced to facilitate different contributions to the title.

#### How to test

1. Start the electron app on Windows or Linux and select the `custom` title bar style in the preferences
2. Assert that the title bar behaves like VSCode when resizing the window.
3. Open a workspace and start opening files or other widgets which appear in the main area. The currently opened workspace and the currently selected main area widget appear in the title bar.
4. Closing all workspaces/main area widgets removes these contributions from the title bar.
5. Assert that these changes also work correctly when using the `native` title bar style and in the browser.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
I haven't looked at the code yet but I believe by reading the description it may fix https://github.com/eclipse-theia/theia/issues/7076.

File: packages/core/src/browser/window/window-title-service.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// *****************************************************************************
+
+import { injectable } from 'inversify';
+import { Emitter, Event } from '../../common/event';
+
+export interface WindowTitlePart {
+    id?: string
+    value?: string
+    priority?: number
+    event?: Event<string | undefined>
```
Comment (colin-grant-work) [inline packages/core/src/browser/window/window-title-service.ts:24]: No strong feelings here, but in `TabBarToolbarItem`s, the corresponding functionality is under a field `onDidChange`.

**Review by colin-grant-work — COMMENTED** [review]
Functionally, looking good, and only minor comments on the code.

File: packages/core/src/browser/window/window-title-service.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            this.titleParts.splice(index, 1);
+            this.updateTitle();
+        }
+    }
+
+    protected updateTitle(): void {
+        const sortedParts = this.titleParts.sort((a, b) => (b.priority ?? 100) - (a.priority ?? 100));
+        this._title = sortedParts.map(e => e.value).filter(e => e).join(' - ');
+        this.onDidChangeTitleEmitter.fire(this._title);
+    }
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/core/src/browser/window/window-title-service.ts:66]: Ah, I didn't know about this. I think that's a better approach than the title parts. I'll rewrite the feature to accommodate for it :)

**Comment by msujew** [comment]
@colin-grant-work I refactored the feature to use the `window.title` preference. Care to take another look?

File: packages/core/src/browser/window/window-title-service.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                const label = `$\{${key}\}`;
+                const value = this.titleParts.get(key) ?? '';
+                title = title.replace(label, value);
+            }
+            const separatedTitle = title.split('${separator}').filter(e => e.length > 0);
+            this._title = separatedTitle.join(this.separator);
+        }
+        if (this.titleParts.get('developmentHost')) {
+            this._title = this.titleParts.get('developmentHost') + this.separator + this._title;
+        }
```
Comment (colin-grant-work) [inline packages/core/src/browser/window/window-title-service.ts:95]:
```suggestion
        const developmentHost = this.titleParts.get('developmentHost');
        if (developmentHost) {
            this._title = developmentHost + this.separator + this._title;
        }
```

File: packages/core/src/browser/window/window-title-service.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    'activeFolderLong': true,
+    'folderName': true,
+    'folderPath': true,
+    'rootName': true,
+    'rootPath': true,
+    'appName': true,
+    'remoteName': true,
+    'dirty': true,
+    'developmentHost': true
+};
```
Comment (colin-grant-work) [inline packages/core/src/browser/window/window-title-service.ts:36]: I don't mean to make life difficult, but would it be possible to get the best of your original implementation (dynamic contributions) and support (most of) VSCode's options? I can imagine downstream IDE's being interested in adding contributions here, maybe variations on 'project type' fields or other custom status markers?

File: packages/core/src/browser/window/window-title-updater.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    protected updateTitleWidget(widget?: Widget): void {
+        let activeEditorLong: string | undefined;
+        let activeEditorMedium: string | undefined;
+        let activeEditorShort: string | undefined;
+        let activeFolderLong: string | undefined;
+        let activeFolderMedium: string | undefined;
+        let activeFolderShort: string | undefined;
+        if (Navigatable.is(widget) && widget.getResourceUri()) {
+            const uri = widget.getResourceUri()!;
```
Comment (colin-grant-work) [inline packages/core/src/browser/window/window-title-updater.ts:47]:
```suggestion
        const uri = NavigatableWidget.getUri(widget);
        if (uri) {
```

File: packages/core/src/browser/window/window-title-service.ts (hunk 4 above)
Comment (msujew) [reply packages/core/src/browser/window/window-title-service.ts]: I've made the `WindowTitleService` a bit more lenient in what it accepts. Downstream apps can now define custom parts 👍

**Review by colin-grant-work — COMMENTED** [review]
This looks good to me. There are just two things I wanted to check on, and then a couple of odd behaviors associated with them.

- It looks like for the `activeEditor` fields, all widgets are eligible, main area and otherwise. Is that intentional, or did you want to restrict to editors or main-area widgets (this is trickier for us than VSCode, so I'm happy with whatever policy, as long as the behavior is as expected).
- In Electron, when the window is no longer the active window (e.g. alt-tab to a browser window), `activeEditor` fields disappear - I guess that the `activeWidget` is set to `undefined` in that case.
- Because `activeEditor` fields reflect all widgets, there are a number of quirks.
    - In the Theia workspace, the title for the navigator tree is `theia` so you can end up with this:

`${activeEditorLong}${separator}${activeEditorShort}${separator}${rootName}`  
image

**Comment by msujew** [comment]
@colin-grant-work Thanks for the feedback! I refactored some stuff, so that only the `main` area is considered for title changes. This should result in less unexpected behavior. Also correctly handles the case when the electron window loses focus.

**Review by vince-fugnitto — COMMENTED** [review]
I haven't tested thoroughly yet but I noticed that the menu does not update properly when a view is opened and closed.

image

For example, in this use-case I:
- opened the preferences-view
- the menu displays `Preferences - theia - Theia Electron Example`
- closed the preferences-view
- the menu persists with `Preferences - theia - Theia Electron Example`

**Comment by msujew** [comment]
@vince-fugnitto It seems like `onDidRemoveWidget` fires a few frames before the last widget of the main area is actually removed, which is why it stays in the title. I added special handling for that case 👍

**Comment by colin-grant-work** [comment]
Another odd sequence:

 - Open Preferences -> Preferences in title 👍 
 - Open Explore -> Preferences in title 👍 
 - Single-click a file to open preview -> Preferences still in title ❓ 
 - Change OS-window focus (for example, switch to a browser to write this comment :-)) -> Editor in title 👍 ❓

**Comment by colin-grant-work** [comment]
> Instead of trying to rely on unreliable events, I've changed the code to query the current main area widget every 33ms.

I'll take a look, but I'm a little nervous about this premise. In general, the cure for unreliable events shouldn't be timeouts but fixing the events so that they're reliable.

**Comment by msujew** [comment]
@colin-grant-work I looked into it again (and rebased everything) and I might need to clarify my previous statement:

The events that are currently available on the `ApplicationShell` (`onDidAddWidget`/`onDidRemoveWidget`/`onDidChangeActiveWidget`) are reliable in the way they are usually used. I.e. identifying if anything has changed. However, we don't try to identify all changes, but only the ones that are relevant to the `main` area of the app. These events don't allow us to filter on that (as in: we can't identify whether the widget which was changed/removed was previously in the main area) and so they are not usable for this feature. The only reliable way to actually identify what the current main area widget is, is to use `setInterval` in a tight loop.

So with that said, do you mind giving this another review?

**Comment by colin-grant-work** [comment]
@msujew, I will take a look at this by the end of the week.

**Comment by colin-grant-work** [comment]
> The only reliable way to actually identify what the current main area widget is, is to use `setInterval` in a tight loop.

I can't embrace this approach. I don't really think it's going to be a significant performance drain - the execution should be quite quick - but if it's true that that's the only option, it does mean that our event system is not as informative as it needs to be for quite a straightforward use case. I will do some looking at whether I see an alternative.

**Comment by colin-grant-work** [comment]
It looks like code like [this](https://github.com/colin-grant-work/theia/commit/d686c0ac250c5132f8b87a7f9d7a904aa9efc503) handles the edge cases pretty well. See what you think and use it if it seems useful.

**Comment by msujew** [comment]
@colin-grant-work Thanks a lot for the clue to look into the dock panel events. That works perfectly even without the interval 👍 

Note that I still had to use the `Saveable.isDirty` check when the main widget changes. Without it, the title wouldn't pick up on switches to an already dirty widget.

**Review by colin-grant-work — APPROVED** [review]
Consistently updates the title according to the user's preference. 👍

**Pull request merged**